core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -577,13 +582,17 @@ private[deploy] class Worker(
     val shouldCleanup = finishedApps.contains(id) && !executors.values.exists(_.appId == id)
     if (shouldCleanup) {
       finishedApps -= id
-      appDirectories.remove(id).foreach { dirList =>
-        logInfo(s"Cleaning up local directories for application $id")
-        dirList.foreach { dir =>
-          Utils.deleteRecursively(new File(dir))
+      cleanupApplicationThreadExecutor.submit(new Runnable {
```

Comment:
I agree with you, but I am worried that cleaning up the workDir and application all cost mush time.